src/governance.cpp

```diff
@@ -927,13 +943,35 @@ void CGovernanceManager::CheckMasternodeOrphanObjects()
     fRateChecksEnabled = true;
 }
 
-void CGovernanceManager::RequestGovernanceObject(CNode* pfrom, const uint256& nHash)
+void CGovernanceManager::RequestGovernanceObject(CNode* pfrom, const uint256& nHash, bool fUseFilter)
 {
     if(!pfrom) {
         return;
     }
 
-    pfrom->PushMessage(NetMsgType::MNGOVERNANCESYNC, nHash);
+    if(pfrom->nVersion < GOVERNANCE_FILTER_PROTO_VERSION) {
+        pfrom->PushMessage(NetMsgType::MNGOVERNANCESYNC, nHash);
+        return;
+    }
+
+    CBloomFilter filter(GOVERNANCE_FILTER_ELEMENTS, GOVERNANCE_FILTER_FP_RATE, GetRandInt(999999), BLOOM_UPDATE_ALL);
```

Comment:
We can use empty filter when `fUseFilter == false` or `pObj == NULL` (1KB vs 36KB). For other cases we can probably use dynamic bloom filter size based on `N of masternodes` and `max N of signals per obj`, which should make more sense on testnet and still fits mainnet.